// @devvit/protos/types/devvit/orchestrator/orchestrator.js
/**
 * #orchestrator.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from "protobufjs/minimal.js";
import { messageTypeRegistry } from "../../typeRegistry.js";
/**
 * ComputeCluster represents the compute infrastructure cluster where an app
 * installation runs.
 */
export var ComputeCluster;
(function (ComputeCluster) {
    /** COMPUTE_CLUSTER_UNSPECIFIED - Unspecified cluster. The orchestrator treats this as ELYSIUM_GCP. */
    ComputeCluster[ComputeCluster["COMPUTE_CLUSTER_UNSPECIFIED"] = 0] = "COMPUTE_CLUSTER_UNSPECIFIED";
    /** COMPUTE_GO - devvit-compute-go runtime. */
    ComputeCluster[ComputeCluster["COMPUTE_GO"] = 1] = "COMPUTE_GO";
    /** ELYSIUM_GCP - Elysium production runtime, apps on GCP. */
    ComputeCluster[ComputeCluster["ELYSIUM_GCP"] = 2] = "ELYSIUM_GCP";
    /** COMPUTE_GO_STAGING - devvit-compute-go staging runtime. */
    ComputeCluster[ComputeCluster["COMPUTE_GO_STAGING"] = 3] = "COMPUTE_GO_STAGING";
    /** ELYSIUM_GCP_STAGING - Elysium staging runtime, apps on GCP. */
    ComputeCluster[ComputeCluster["ELYSIUM_GCP_STAGING"] = 4] = "ELYSIUM_GCP_STAGING";
    ComputeCluster[ComputeCluster["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(ComputeCluster || (ComputeCluster = {}));
export function computeClusterFromJSON(object) {
    switch (object) {
        case 0:
        case "COMPUTE_CLUSTER_UNSPECIFIED":
            return ComputeCluster.COMPUTE_CLUSTER_UNSPECIFIED;
        case 1:
        case "COMPUTE_GO":
            return ComputeCluster.COMPUTE_GO;
        case 2:
        case "ELYSIUM_GCP":
            return ComputeCluster.ELYSIUM_GCP;
        case 3:
        case "COMPUTE_GO_STAGING":
            return ComputeCluster.COMPUTE_GO_STAGING;
        case 4:
        case "ELYSIUM_GCP_STAGING":
            return ComputeCluster.ELYSIUM_GCP_STAGING;
        case -1:
        case "UNRECOGNIZED":
        default:
            return ComputeCluster.UNRECOGNIZED;
    }
}
export function computeClusterToJSON(object) {
    switch (object) {
        case ComputeCluster.COMPUTE_CLUSTER_UNSPECIFIED:
            return 0;
        case ComputeCluster.COMPUTE_GO:
            return 1;
        case ComputeCluster.ELYSIUM_GCP:
            return 2;
        case ComputeCluster.COMPUTE_GO_STAGING:
            return 3;
        case ComputeCluster.ELYSIUM_GCP_STAGING:
            return 4;
        case ComputeCluster.UNRECOGNIZED:
        default:
            return -1;
    }
}
function createBaseSetDeploymentRequest() {
    return { installationId: "", linkedBundleUrl: undefined, computeCluster: 0 };
}
export const SetDeploymentRequest = {
    $type: "devvit.orchestrator.SetDeploymentRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.installationId !== "") {
            writer.uint32(10).string(message.installationId);
        }
        if (message.linkedBundleUrl !== undefined) {
            writer.uint32(18).string(message.linkedBundleUrl);
        }
        if (message.computeCluster !== 0) {
            writer.uint32(24).int32(message.computeCluster);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSetDeploymentRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.installationId = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.linkedBundleUrl = reader.string();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.computeCluster = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            installationId: isSet(object.installationId) ? globalThis.String(object.installationId) : "",
            linkedBundleUrl: isSet(object.linkedBundleUrl) ? globalThis.String(object.linkedBundleUrl) : undefined,
            computeCluster: isSet(object.computeCluster) ? computeClusterFromJSON(object.computeCluster) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.installationId !== "") {
            obj.installationId = message.installationId;
        }
        if (message.linkedBundleUrl !== undefined) {
            obj.linkedBundleUrl = message.linkedBundleUrl;
        }
        if (message.computeCluster !== 0) {
            obj.computeCluster = computeClusterToJSON(message.computeCluster);
        }
        return obj;
    },
    create(base) {
        return SetDeploymentRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseSetDeploymentRequest();
        message.installationId = object.installationId ?? "";
        message.linkedBundleUrl = object.linkedBundleUrl ?? undefined;
        message.computeCluster = object.computeCluster ?? 0;
        return message;
    },
};
messageTypeRegistry.set(SetDeploymentRequest.$type, SetDeploymentRequest);
function createBaseSetDeploymentResponse() {
    return { url: "" };
}
export const SetDeploymentResponse = {
    $type: "devvit.orchestrator.SetDeploymentResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.url !== "") {
            writer.uint32(10).string(message.url);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSetDeploymentResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.url = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { url: isSet(object.url) ? globalThis.String(object.url) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.url !== "") {
            obj.url = message.url;
        }
        return obj;
    },
    create(base) {
        return SetDeploymentResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseSetDeploymentResponse();
        message.url = object.url ?? "";
        return message;
    },
};
messageTypeRegistry.set(SetDeploymentResponse.$type, SetDeploymentResponse);
function createBaseDeleteDeploymentRequest() {
    return { installationId: "", computeCluster: 0 };
}
export const DeleteDeploymentRequest = {
    $type: "devvit.orchestrator.DeleteDeploymentRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.installationId !== "") {
            writer.uint32(10).string(message.installationId);
        }
        if (message.computeCluster !== 0) {
            writer.uint32(16).int32(message.computeCluster);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDeleteDeploymentRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.installationId = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.computeCluster = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            installationId: isSet(object.installationId) ? globalThis.String(object.installationId) : "",
            computeCluster: isSet(object.computeCluster) ? computeClusterFromJSON(object.computeCluster) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.installationId !== "") {
            obj.installationId = message.installationId;
        }
        if (message.computeCluster !== 0) {
            obj.computeCluster = computeClusterToJSON(message.computeCluster);
        }
        return obj;
    },
    create(base) {
        return DeleteDeploymentRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseDeleteDeploymentRequest();
        message.installationId = object.installationId ?? "";
        message.computeCluster = object.computeCluster ?? 0;
        return message;
    },
};
messageTypeRegistry.set(DeleteDeploymentRequest.$type, DeleteDeploymentRequest);
function createBaseDeleteDeploymentResponse() {
    return {};
}
export const DeleteDeploymentResponse = {
    $type: "devvit.orchestrator.DeleteDeploymentResponse",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDeleteDeploymentResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return DeleteDeploymentResponse.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseDeleteDeploymentResponse();
        return message;
    },
};
messageTypeRegistry.set(DeleteDeploymentResponse.$type, DeleteDeploymentResponse);
export const OrchestratorServiceName = "devvit.orchestrator.Orchestrator";
export class OrchestratorClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || OrchestratorServiceName;
        this.rpc = rpc;
        this.SetDeployment = this.SetDeployment.bind(this);
        this.DeleteDeployment = this.DeleteDeployment.bind(this);
    }
    SetDeployment(request, metadata) {
        const data = SetDeploymentRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "SetDeployment", data, metadata);
        return promise.then((data) => SetDeploymentResponse.decode(_m0.Reader.create(data)));
    }
    DeleteDeployment(request, metadata) {
        const data = DeleteDeploymentRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DeleteDeployment", data, metadata);
        return promise.then((data) => DeleteDeploymentResponse.decode(_m0.Reader.create(data)));
    }
}
export const OrchestratorDefinition = {
    name: "Orchestrator",
    fullName: "devvit.orchestrator.Orchestrator",
    methods: {
        setDeployment: {
            name: "SetDeployment",
            requestType: SetDeploymentRequest,
            requestStream: false,
            responseType: SetDeploymentResponse,
            responseStream: false,
            options: {},
        },
        deleteDeployment: {
            name: "DeleteDeployment",
            requestType: DeleteDeploymentRequest,
            requestStream: false,
            responseType: DeleteDeploymentResponse,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
