// @devvit/protos/types/devvit/plugin/externalendpoints/v1alpha/externalendpoints_internal.twirp-client.js
import { CreateTokenRequest, CreateTokenResponse, ValidateTokenRequest, ValidateTokenResponse, ListTokensRequest, ListTokensResponse, RevokeTokenRequest, RevokeTokenResponse, } from './externalendpoints_internal.js';
export class ExternalEndpointsInternalClientJSON {
    constructor(rpc) {
        this.rpc = rpc;
        this.CreateToken.bind(this);
        this.ValidateToken.bind(this);
        this.ListTokens.bind(this);
        this.RevokeToken.bind(this);
    }
    CreateToken(request) {
        const data = CreateTokenRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal', 'CreateToken', 'application/json', data);
        return promise.then((data) => CreateTokenResponse.fromJSON(data));
    }
    ValidateToken(request) {
        const data = ValidateTokenRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal', 'ValidateToken', 'application/json', data);
        return promise.then((data) => ValidateTokenResponse.fromJSON(data));
    }
    ListTokens(request) {
        const data = ListTokensRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal', 'ListTokens', 'application/json', data);
        return promise.then((data) => ListTokensResponse.fromJSON(data));
    }
    RevokeToken(request) {
        const data = RevokeTokenRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal', 'RevokeToken', 'application/json', data);
        return promise.then((data) => RevokeTokenResponse.fromJSON(data));
    }
}
export class ExternalEndpointsInternalClientProtobuf {
    constructor(rpc) {
        this.rpc = rpc;
        this.CreateToken.bind(this);
        this.ValidateToken.bind(this);
        this.ListTokens.bind(this);
        this.RevokeToken.bind(this);
    }
    CreateToken(request) {
        const data = CreateTokenRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal', 'CreateToken', 'application/protobuf', data);
        return promise.then((data) => CreateTokenResponse.decode(data));
    }
    ValidateToken(request) {
        const data = ValidateTokenRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal', 'ValidateToken', 'application/protobuf', data);
        return promise.then((data) => ValidateTokenResponse.decode(data));
    }
    ListTokens(request) {
        const data = ListTokensRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal', 'ListTokens', 'application/protobuf', data);
        return promise.then((data) => ListTokensResponse.decode(data));
    }
    RevokeToken(request) {
        const data = RevokeTokenRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal', 'RevokeToken', 'application/protobuf', data);
        return promise.then((data) => RevokeTokenResponse.decode(data));
    }
}
