// @devvit/protos/types/devvit/plugin/externalendpoints/v1alpha/externalendpoints.js
/**
 * #externalendpoints.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from "protobufjs/minimal.js";
import { messageTypeRegistry } from "../../../../typeRegistry.js";
/** External endpoint token / endpoint access permission. */
export var Scope;
(function (Scope) {
    Scope[Scope["SCOPE_UNSPECIFIED"] = 0] = "SCOPE_UNSPECIFIED";
    /**
     * SCOPE_INSTALL - Install allows short-lived callback tokens generated by the app in an
     * outgoing `fetch()` wherever the app is installed (a subreddit or user
     * account).
     */
    Scope[Scope["SCOPE_INSTALL"] = 1] = "SCOPE_INSTALL";
    /** SCOPE_GLOBAL - Global allows long-lived app tokens. */
    Scope[Scope["SCOPE_GLOBAL"] = 2] = "SCOPE_GLOBAL";
    Scope[Scope["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(Scope || (Scope = {}));
export function scopeFromJSON(object) {
    switch (object) {
        case 0:
        case "SCOPE_UNSPECIFIED":
            return Scope.SCOPE_UNSPECIFIED;
        case 1:
        case "SCOPE_INSTALL":
            return Scope.SCOPE_INSTALL;
        case 2:
        case "SCOPE_GLOBAL":
            return Scope.SCOPE_GLOBAL;
        case -1:
        case "UNRECOGNIZED":
        default:
            return Scope.UNRECOGNIZED;
    }
}
export function scopeToJSON(object) {
    switch (object) {
        case Scope.SCOPE_UNSPECIFIED:
            return 0;
        case Scope.SCOPE_INSTALL:
            return 1;
        case Scope.SCOPE_GLOBAL:
            return 2;
        case Scope.UNRECOGNIZED:
        default:
            return -1;
    }
}
function createBaseGetCallbackUrlRequest() {
    return { callbackHandlerPath: "" };
}
export const GetCallbackUrlRequest = {
    $type: "devvit.plugin.externalendpoints.v1alpha.GetCallbackUrlRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.callbackHandlerPath !== "") {
            writer.uint32(10).string(message.callbackHandlerPath);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetCallbackUrlRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.callbackHandlerPath = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            callbackHandlerPath: isSet(object.callbackHandlerPath) ? globalThis.String(object.callbackHandlerPath) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.callbackHandlerPath !== "") {
            obj.callbackHandlerPath = message.callbackHandlerPath;
        }
        return obj;
    },
    create(base) {
        return GetCallbackUrlRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetCallbackUrlRequest();
        message.callbackHandlerPath = object.callbackHandlerPath ?? "";
        return message;
    },
};
messageTypeRegistry.set(GetCallbackUrlRequest.$type, GetCallbackUrlRequest);
function createBaseGetCallbackUrlResponse() {
    return { callbackUrl: "" };
}
export const GetCallbackUrlResponse = {
    $type: "devvit.plugin.externalendpoints.v1alpha.GetCallbackUrlResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.callbackUrl !== "") {
            writer.uint32(10).string(message.callbackUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetCallbackUrlResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.callbackUrl = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { callbackUrl: isSet(object.callbackUrl) ? globalThis.String(object.callbackUrl) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.callbackUrl !== "") {
            obj.callbackUrl = message.callbackUrl;
        }
        return obj;
    },
    create(base) {
        return GetCallbackUrlResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetCallbackUrlResponse();
        message.callbackUrl = object.callbackUrl ?? "";
        return message;
    },
};
messageTypeRegistry.set(GetCallbackUrlResponse.$type, GetCallbackUrlResponse);
export const ExternalEndpointsServiceName = "devvit.plugin.externalendpoints.v1alpha.ExternalEndpoints";
export class ExternalEndpointsClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || ExternalEndpointsServiceName;
        this.rpc = rpc;
        this.GetCallbackUrl = this.GetCallbackUrl.bind(this);
    }
    GetCallbackUrl(request, metadata) {
        const data = GetCallbackUrlRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetCallbackUrl", data, metadata);
        return promise.then((data) => GetCallbackUrlResponse.decode(_m0.Reader.create(data)));
    }
}
export const ExternalEndpointsDefinition = {
    name: "ExternalEndpoints",
    fullName: "devvit.plugin.externalendpoints.v1alpha.ExternalEndpoints",
    methods: {
        /** Get a callback URL with a token that can be used to validate the request */
        getCallbackUrl: {
            name: "GetCallbackUrl",
            requestType: GetCallbackUrlRequest,
            requestStream: false,
            responseType: GetCallbackUrlResponse,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
