// @devvit/protos/types/devvit/plugin/externalendpoints/v1alpha/externalendpoints.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { GetCallbackUrlRequest, GetCallbackUrlResponse, } from './externalendpoints.js';
export var ExternalEndpointsMethod;
(function (ExternalEndpointsMethod) {
    ExternalEndpointsMethod["GetCallbackUrl"] = "GetCallbackUrl";
})(ExternalEndpointsMethod || (ExternalEndpointsMethod = {}));
export const ExternalEndpointsMethodList = [
    ExternalEndpointsMethod.GetCallbackUrl,
];
export function createExternalEndpointsServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.externalendpoints.v1alpha',
        serviceName: 'ExternalEndpoints',
        methodList: ExternalEndpointsMethodList,
        matchRoute: matchExternalEndpointsRoute,
    });
}
function matchExternalEndpointsRoute(method, events) {
    switch (method) {
        case 'GetCallbackUrl':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetCallbackUrl' };
                await events.onMatch(ctx);
                return handleExternalEndpointsGetCallbackUrlRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleExternalEndpointsGetCallbackUrlRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleExternalEndpointsGetCallbackUrlJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleExternalEndpointsGetCallbackUrlProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleExternalEndpointsGetCallbackUrlJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetCallbackUrlRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetCallbackUrl(ctx, inputReq);
        });
    }
    else {
        response = await service.GetCallbackUrl(ctx, request);
    }
    return JSON.stringify(GetCallbackUrlResponse.toJSON(response));
}
async function handleExternalEndpointsGetCallbackUrlProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetCallbackUrlRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetCallbackUrl(ctx, inputReq);
        });
    }
    else {
        response = await service.GetCallbackUrl(ctx, request);
    }
    return Buffer.from(GetCallbackUrlResponse.encode(response).finish());
}
