// @devvit/protos/types/devvit/plugin/externalendpoints/v1alpha/externalendpoints_internal.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { CreateTokenRequest, CreateTokenResponse, ValidateTokenRequest, ValidateTokenResponse, ListTokensRequest, ListTokensResponse, RevokeTokenRequest, RevokeTokenResponse, } from './externalendpoints_internal.js';
export var ExternalEndpointsInternalMethod;
(function (ExternalEndpointsInternalMethod) {
    ExternalEndpointsInternalMethod["CreateToken"] = "CreateToken";
    ExternalEndpointsInternalMethod["ValidateToken"] = "ValidateToken";
    ExternalEndpointsInternalMethod["ListTokens"] = "ListTokens";
    ExternalEndpointsInternalMethod["RevokeToken"] = "RevokeToken";
})(ExternalEndpointsInternalMethod || (ExternalEndpointsInternalMethod = {}));
export const ExternalEndpointsInternalMethodList = [
    ExternalEndpointsInternalMethod.CreateToken,
    ExternalEndpointsInternalMethod.ValidateToken,
    ExternalEndpointsInternalMethod.ListTokens,
    ExternalEndpointsInternalMethod.RevokeToken,
];
export function createExternalEndpointsInternalServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.externalendpoints.v1alpha',
        serviceName: 'ExternalEndpointsInternal',
        methodList: ExternalEndpointsInternalMethodList,
        matchRoute: matchExternalEndpointsInternalRoute,
    });
}
function matchExternalEndpointsInternalRoute(method, events) {
    switch (method) {
        case 'CreateToken':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'CreateToken' };
                await events.onMatch(ctx);
                return handleExternalEndpointsInternalCreateTokenRequest(ctx, service, data, interceptors);
            };
        case 'ValidateToken':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'ValidateToken' };
                await events.onMatch(ctx);
                return handleExternalEndpointsInternalValidateTokenRequest(ctx, service, data, interceptors);
            };
        case 'ListTokens':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'ListTokens' };
                await events.onMatch(ctx);
                return handleExternalEndpointsInternalListTokensRequest(ctx, service, data, interceptors);
            };
        case 'RevokeToken':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'RevokeToken' };
                await events.onMatch(ctx);
                return handleExternalEndpointsInternalRevokeTokenRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleExternalEndpointsInternalCreateTokenRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleExternalEndpointsInternalCreateTokenJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleExternalEndpointsInternalCreateTokenProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleExternalEndpointsInternalValidateTokenRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleExternalEndpointsInternalValidateTokenJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleExternalEndpointsInternalValidateTokenProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleExternalEndpointsInternalListTokensRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleExternalEndpointsInternalListTokensJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleExternalEndpointsInternalListTokensProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleExternalEndpointsInternalRevokeTokenRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleExternalEndpointsInternalRevokeTokenJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleExternalEndpointsInternalRevokeTokenProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleExternalEndpointsInternalCreateTokenJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = CreateTokenRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.CreateToken(ctx, inputReq);
        });
    }
    else {
        response = await service.CreateToken(ctx, request);
    }
    return JSON.stringify(CreateTokenResponse.toJSON(response));
}
async function handleExternalEndpointsInternalValidateTokenJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = ValidateTokenRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ValidateToken(ctx, inputReq);
        });
    }
    else {
        response = await service.ValidateToken(ctx, request);
    }
    return JSON.stringify(ValidateTokenResponse.toJSON(response));
}
async function handleExternalEndpointsInternalListTokensJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = ListTokensRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ListTokens(ctx, inputReq);
        });
    }
    else {
        response = await service.ListTokens(ctx, request);
    }
    return JSON.stringify(ListTokensResponse.toJSON(response));
}
async function handleExternalEndpointsInternalRevokeTokenJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = RevokeTokenRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RevokeToken(ctx, inputReq);
        });
    }
    else {
        response = await service.RevokeToken(ctx, request);
    }
    return JSON.stringify(RevokeTokenResponse.toJSON(response));
}
async function handleExternalEndpointsInternalCreateTokenProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = CreateTokenRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.CreateToken(ctx, inputReq);
        });
    }
    else {
        response = await service.CreateToken(ctx, request);
    }
    return Buffer.from(CreateTokenResponse.encode(response).finish());
}
async function handleExternalEndpointsInternalValidateTokenProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = ValidateTokenRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ValidateToken(ctx, inputReq);
        });
    }
    else {
        response = await service.ValidateToken(ctx, request);
    }
    return Buffer.from(ValidateTokenResponse.encode(response).finish());
}
async function handleExternalEndpointsInternalListTokensProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = ListTokensRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ListTokens(ctx, inputReq);
        });
    }
    else {
        response = await service.ListTokens(ctx, request);
    }
    return Buffer.from(ListTokensResponse.encode(response).finish());
}
async function handleExternalEndpointsInternalRevokeTokenProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = RevokeTokenRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RevokeToken(ctx, inputReq);
        });
    }
    else {
        response = await service.RevokeToken(ctx, request);
    }
    return Buffer.from(RevokeTokenResponse.encode(response).finish());
}
