// @devvit/protos/types/devvit/plugin/externalendpoints/v1alpha/externalendpoints_internal.js
/**
 * #externalendpoints_internal.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from "protobufjs/minimal.js";
import { Timestamp } from "../../../../google/protobuf/timestamp.js";
import { messageTypeRegistry } from "../../../../typeRegistry.js";
function createBaseToken() {
    return { id: "", name: "", value: undefined, createdAt: undefined, createdBy: "", revoked: false };
}
export const Token = {
    $type: "devvit.plugin.externalendpoints.v1alpha.Token",
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.name !== "") {
            writer.uint32(18).string(message.name);
        }
        if (message.value !== undefined) {
            writer.uint32(26).string(message.value);
        }
        if (message.createdAt !== undefined) {
            Timestamp.encode(toTimestamp(message.createdAt), writer.uint32(34).fork()).ldelim();
        }
        if (message.createdBy !== "") {
            writer.uint32(42).string(message.createdBy);
        }
        if (message.revoked !== false) {
            writer.uint32(48).bool(message.revoked);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseToken();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.createdAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.createdBy = reader.string();
                    continue;
                case 6:
                    if (tag !== 48) {
                        break;
                    }
                    message.revoked = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.String(object.id) : "",
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : undefined,
            createdAt: isSet(object.createdAt) ? fromJsonTimestamp(object.createdAt) : undefined,
            createdBy: isSet(object.createdBy) ? globalThis.String(object.createdBy) : "",
            revoked: isSet(object.revoked) ? globalThis.Boolean(object.revoked) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== "") {
            obj.id = message.id;
        }
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.value !== undefined) {
            obj.value = message.value;
        }
        if (message.createdAt !== undefined) {
            obj.createdAt = message.createdAt.toISOString();
        }
        if (message.createdBy !== "") {
            obj.createdBy = message.createdBy;
        }
        if (message.revoked !== false) {
            obj.revoked = message.revoked;
        }
        return obj;
    },
    create(base) {
        return Token.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseToken();
        message.id = object.id ?? "";
        message.name = object.name ?? "";
        message.value = object.value ?? undefined;
        message.createdAt = object.createdAt ?? undefined;
        message.createdBy = object.createdBy ?? "";
        message.revoked = object.revoked ?? false;
        return message;
    },
};
messageTypeRegistry.set(Token.$type, Token);
function createBaseCreateTokenRequest() {
    return { tokenName: "" };
}
export const CreateTokenRequest = {
    $type: "devvit.plugin.externalendpoints.v1alpha.CreateTokenRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.tokenName !== "") {
            writer.uint32(10).string(message.tokenName);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateTokenRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.tokenName = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { tokenName: isSet(object.tokenName) ? globalThis.String(object.tokenName) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.tokenName !== "") {
            obj.tokenName = message.tokenName;
        }
        return obj;
    },
    create(base) {
        return CreateTokenRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCreateTokenRequest();
        message.tokenName = object.tokenName ?? "";
        return message;
    },
};
messageTypeRegistry.set(CreateTokenRequest.$type, CreateTokenRequest);
function createBaseCreateTokenResponse() {
    return { token: undefined };
}
export const CreateTokenResponse = {
    $type: "devvit.plugin.externalendpoints.v1alpha.CreateTokenResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.token !== undefined) {
            Token.encode(message.token, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateTokenResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.token = Token.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { token: isSet(object.token) ? Token.fromJSON(object.token) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.token !== undefined) {
            obj.token = Token.toJSON(message.token);
        }
        return obj;
    },
    create(base) {
        return CreateTokenResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseCreateTokenResponse();
        message.token = (object.token !== undefined && object.token !== null) ? Token.fromPartial(object.token) : undefined;
        return message;
    },
};
messageTypeRegistry.set(CreateTokenResponse.$type, CreateTokenResponse);
function createBaseValidateTokenRequest() {
    return { tokenValue: "" };
}
export const ValidateTokenRequest = {
    $type: "devvit.plugin.externalendpoints.v1alpha.ValidateTokenRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.tokenValue !== "") {
            writer.uint32(10).string(message.tokenValue);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseValidateTokenRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.tokenValue = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { tokenValue: isSet(object.tokenValue) ? globalThis.String(object.tokenValue) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.tokenValue !== "") {
            obj.tokenValue = message.tokenValue;
        }
        return obj;
    },
    create(base) {
        return ValidateTokenRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseValidateTokenRequest();
        message.tokenValue = object.tokenValue ?? "";
        return message;
    },
};
messageTypeRegistry.set(ValidateTokenRequest.$type, ValidateTokenRequest);
function createBaseValidateTokenResponse() {
    return { valid: false };
}
export const ValidateTokenResponse = {
    $type: "devvit.plugin.externalendpoints.v1alpha.ValidateTokenResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.valid !== false) {
            writer.uint32(8).bool(message.valid);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseValidateTokenResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.valid = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { valid: isSet(object.valid) ? globalThis.Boolean(object.valid) : false };
    },
    toJSON(message) {
        const obj = {};
        if (message.valid !== false) {
            obj.valid = message.valid;
        }
        return obj;
    },
    create(base) {
        return ValidateTokenResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseValidateTokenResponse();
        message.valid = object.valid ?? false;
        return message;
    },
};
messageTypeRegistry.set(ValidateTokenResponse.$type, ValidateTokenResponse);
function createBaseListTokensRequest() {
    return { appId: "" };
}
export const ListTokensRequest = {
    $type: "devvit.plugin.externalendpoints.v1alpha.ListTokensRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.appId !== "") {
            writer.uint32(10).string(message.appId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListTokensRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.appId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { appId: isSet(object.appId) ? globalThis.String(object.appId) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.appId !== "") {
            obj.appId = message.appId;
        }
        return obj;
    },
    create(base) {
        return ListTokensRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseListTokensRequest();
        message.appId = object.appId ?? "";
        return message;
    },
};
messageTypeRegistry.set(ListTokensRequest.$type, ListTokensRequest);
function createBaseListTokensResponse() {
    return { tokens: [] };
}
export const ListTokensResponse = {
    $type: "devvit.plugin.externalendpoints.v1alpha.ListTokensResponse",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.tokens) {
            Token.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListTokensResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.tokens.push(Token.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { tokens: globalThis.Array.isArray(object?.tokens) ? object.tokens.map((e) => Token.fromJSON(e)) : [] };
    },
    toJSON(message) {
        const obj = {};
        if (message.tokens?.length) {
            obj.tokens = message.tokens.map((e) => Token.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return ListTokensResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseListTokensResponse();
        message.tokens = object.tokens?.map((e) => Token.fromPartial(e)) || [];
        return message;
    },
};
messageTypeRegistry.set(ListTokensResponse.$type, ListTokensResponse);
function createBaseRevokeTokenRequest() {
    return { tokenId: "" };
}
export const RevokeTokenRequest = {
    $type: "devvit.plugin.externalendpoints.v1alpha.RevokeTokenRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.tokenId !== "") {
            writer.uint32(10).string(message.tokenId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRevokeTokenRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.tokenId = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { tokenId: isSet(object.tokenId) ? globalThis.String(object.tokenId) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.tokenId !== "") {
            obj.tokenId = message.tokenId;
        }
        return obj;
    },
    create(base) {
        return RevokeTokenRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRevokeTokenRequest();
        message.tokenId = object.tokenId ?? "";
        return message;
    },
};
messageTypeRegistry.set(RevokeTokenRequest.$type, RevokeTokenRequest);
function createBaseRevokeTokenResponse() {
    return { revoked: false };
}
export const RevokeTokenResponse = {
    $type: "devvit.plugin.externalendpoints.v1alpha.RevokeTokenResponse",
    encode(message, writer = _m0.Writer.create()) {
        if (message.revoked !== false) {
            writer.uint32(8).bool(message.revoked);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRevokeTokenResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.revoked = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { revoked: isSet(object.revoked) ? globalThis.Boolean(object.revoked) : false };
    },
    toJSON(message) {
        const obj = {};
        if (message.revoked !== false) {
            obj.revoked = message.revoked;
        }
        return obj;
    },
    create(base) {
        return RevokeTokenResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRevokeTokenResponse();
        message.revoked = object.revoked ?? false;
        return message;
    },
};
messageTypeRegistry.set(RevokeTokenResponse.$type, RevokeTokenResponse);
export const ExternalEndpointsInternalServiceName = "devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal";
export class ExternalEndpointsInternalClientImpl {
    constructor(rpc, opts) {
        this.service = opts?.service || ExternalEndpointsInternalServiceName;
        this.rpc = rpc;
        this.CreateToken = this.CreateToken.bind(this);
        this.ValidateToken = this.ValidateToken.bind(this);
        this.ListTokens = this.ListTokens.bind(this);
        this.RevokeToken = this.RevokeToken.bind(this);
    }
    CreateToken(request, metadata) {
        const data = CreateTokenRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CreateToken", data, metadata);
        return promise.then((data) => CreateTokenResponse.decode(_m0.Reader.create(data)));
    }
    ValidateToken(request, metadata) {
        const data = ValidateTokenRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "ValidateToken", data, metadata);
        return promise.then((data) => ValidateTokenResponse.decode(_m0.Reader.create(data)));
    }
    ListTokens(request, metadata) {
        const data = ListTokensRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "ListTokens", data, metadata);
        return promise.then((data) => ListTokensResponse.decode(_m0.Reader.create(data)));
    }
    RevokeToken(request, metadata) {
        const data = RevokeTokenRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "RevokeToken", data, metadata);
        return promise.then((data) => RevokeTokenResponse.decode(_m0.Reader.create(data)));
    }
}
export const ExternalEndpointsInternalDefinition = {
    name: "ExternalEndpointsInternal",
    fullName: "devvit.plugin.externalendpoints.v1alpha.ExternalEndpointsInternal",
    methods: {
        /** Create a long-term token that can be used to validate requests from external sources */
        createToken: {
            name: "CreateToken",
            requestType: CreateTokenRequest,
            requestStream: false,
            responseType: CreateTokenResponse,
            responseStream: false,
            options: {},
        },
        /** Validate a token (both short and long-term) */
        validateToken: {
            name: "ValidateToken",
            requestType: ValidateTokenRequest,
            requestStream: false,
            responseType: ValidateTokenResponse,
            responseStream: false,
            options: {},
        },
        /** List all tokens for an app */
        listTokens: {
            name: "ListTokens",
            requestType: ListTokensRequest,
            requestStream: false,
            responseType: ListTokensResponse,
            responseStream: false,
            options: {},
        },
        /** Revoke a token */
        revokeToken: {
            name: "RevokeToken",
            requestType: RevokeTokenRequest,
            requestStream: false,
            responseType: RevokeTokenResponse,
            responseStream: false,
            options: {},
        },
    },
};
function toTimestamp(date) {
    const seconds = Math.trunc(date.getTime() / 1000);
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = (t.seconds || 0) * 1000;
    millis += (t.nanos || 0) / 1000000;
    return new globalThis.Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof globalThis.Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new globalThis.Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
