// @devvit/protos/json/devvit/orchestrator/orchestrator.js
/**
 * #orchestrator.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
/**
 * ComputeCluster represents the compute infrastructure cluster where an app
 * installation runs.
 */
export var ComputeCluster;
(function (ComputeCluster) {
    /** COMPUTE_CLUSTER_UNSPECIFIED - Unspecified cluster. The orchestrator treats this as ELYSIUM_GCP. */
    ComputeCluster[ComputeCluster["COMPUTE_CLUSTER_UNSPECIFIED"] = 0] = "COMPUTE_CLUSTER_UNSPECIFIED";
    /** COMPUTE_GO - devvit-compute-go runtime. */
    ComputeCluster[ComputeCluster["COMPUTE_GO"] = 1] = "COMPUTE_GO";
    /** ELYSIUM_GCP - Elysium production runtime, apps on GCP. */
    ComputeCluster[ComputeCluster["ELYSIUM_GCP"] = 2] = "ELYSIUM_GCP";
    /** COMPUTE_GO_STAGING - devvit-compute-go staging runtime. */
    ComputeCluster[ComputeCluster["COMPUTE_GO_STAGING"] = 3] = "COMPUTE_GO_STAGING";
    /** ELYSIUM_GCP_STAGING - Elysium staging runtime, apps on GCP. */
    ComputeCluster[ComputeCluster["ELYSIUM_GCP_STAGING"] = 4] = "ELYSIUM_GCP_STAGING";
    ComputeCluster[ComputeCluster["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(ComputeCluster || (ComputeCluster = {}));
